src/lib/formats/fdos_dsk.cpp

```diff
@@ -0,0 +1,185 @@
+ *
+ * Disk Format::
+ *
+ * Operating System:: Tracks 0-1
+ * Directory:: Track 2
+ * Data:: Tracks 3-34
+ *
+ *
+ * Note: Due to hardware differences between the original DC-1 (WD1771)
+ * controller for which FDOS was designed and DC-4 (WD1797) or DC-5 (WD2797)
+ * Controllers (DC-5 is Emulated in MAME), the machine can only be booted with
+ * a specific disk image from deramp.com (Mike Douglas) at the following
+ * location.  The other disks available in the same directory can then be read
+ * once FDOS is loaded. This disk image also works on real hardware with a DC-4
+ * or PT FD-2A controllers.
+ *
+ * https://deramp.com/downloads/swtpc/software/FDOS/Disk%20Images/FDOSMPS.DSK
+ *
+ * This disk contains the all of these required patches:
+ *
+ * https://deramp.com/downloads/swtpc/software/FDOS/Disk%20Images/Patches/
+ */
+
+#include "fdos_dsk.h"
+
+#include "imageutl.h"
+
+#include "ioprocs.h"
+
+namespace
+{
+	class fdos_formats : public wd177x_format
+	{
+	public:
+		struct dirent_entry_fdos
+		{
+			char filename[8]{};
+			char password[8]{};
+			uint8_t start_track = 0;
+			uint8_t start_sector = 0;
+			uint8_t num_sectors[2] = {};
+			uint8_t file_type = 0;
+			uint8_t start_addr[2] = {};
+			uint8_t end_addr[2] = {};
+			uint8_t exec_addr[2] = {};
+			uint8_t basic_high_line[2] = {};
+			uint8_t spares[3] = {};
+		};
+
+		static const format formats[];
+	};
+}
+
+fdos_format::fdos_format() : wd177x_format(fdos_formats::formats)
+{
+}
+
+const char *fdos_format::name() const
+{
+	return "fdos";
+}
+
+const char *fdos_format::description() const
+{
+	return "FDOS compatible disk image";
+}
+
+const char *fdos_format::extensions() const
+{
+	return "dsk";
+}
+
+int fdos_format::identify(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants) const
+{
+	int type = find_size(io, form_factor, variants);
+
+	if (type != -1)
+		return FIFID_SIZE;
+	return 0;
+}
+
+
+int fdos_format::find_size(util::random_read &io, uint32_t form_factor, const std::vector<uint32_t> &variants) const
+{
+	uint64_t size;
+	if (io.length(size))
+		return -1;
+
+	uint8_t boot0[256];
+	fdos_formats::dirent_entry_fdos info;
+	size_t actual;
+	std::error_condition ec;
+
+	for (int i=0; fdos_formats::formats[i].form_factor; i++) {
+		const format &f = fdos_formats::formats[i];
+		
+		// Format Check
+		// Check byte 0 and byte 3 of Track 0 Sector 0
+		// 00320 2400 BD 240C START  JSR   BOOT
+		// 00330 2403 DE OB   RESTRT LDX   PROGX
+		// Should be  $BD and $DE respectively
+		// There could be additional variations
+		ec = io.read_at(0, &boot0, f.sector_base_size, actual);
+		if (ec || actual != f.sector_base_size)
+			return -1;
+		if (boot0[0] != 0xbd && boot0[3] != 0xde)
+			continue;
+		
+		LOG_FORMATS("FDOS floppy dsk: size %d bytes, %d total sectors, %d remaining bytes, expected form factor %x\n",
+				(uint32_t)size,
+				(uint32_t)size / f.sector_base_size,
+				(uint32_t)size % f.sector_base_size,
+				form_factor);
```

Comment:
You shouldn’t need to to cast the integers down to print them – `osd_printf_info` and friends are type-safe and handle 64-bit integers fine.